src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -337,16 +337,18 @@ void SolverInterfaceImpl::initialize()
   }
 
   // Initialize coupling state, overwrite these values for restart
-  double time       = 0.0;
-  int    timeWindow = 1;
+  const double time         = 0.0;
+  const int    timeWindow   = 1;
+  const double relativeTime = time::Storage::WINDOW_START;
 
-  for (auto &context : _accessor->readDataContexts()) {
-    context.initializeWaveform();
-  }
   _meshLock.lockAll();
 
+  for (auto &context : _accessor->writeDataContexts()) {
+    context.storeBufferedData(relativeTime);
+  }
+
+  mapWrittenData(); // always perform mapping here to avoid need for workarounds during initialization.
   if (_couplingScheme->sendsInitializedData()) {
-    mapWrittenData();
     performDataActions({action::Action::WRITE_MAPPING_POST}, 0.0);
   }
```

Comment:
I mean that we would need to call `checkDataHasBeenReceived`, even if no data has been received. Main reason: We now have to perform the mapping in initialize to make sure that all Storages of `_fromData` and `_toData` are initialized (potentially zero). But I guess we do not really need this comment.

Currently we still only call the write action if data is actually sent, while the mapping is always executed. But this should improve (see comment).

```
mapWrittenData();
if (_couplingScheme->sendsInitializedData()) { // @todo try to remove this check as well. Currently problematic, because of Integration/Serial/MultipleMappings/MultipleWriteToMappings (During initialization only data at WINDOW_START available. Should improve, if we apply actions to all samples in storage)
  performDataActions({action::Action::WRITE_MAPPING_POST}, 0.0);
}
```

Reply:
This check is not required anymore. Removed it in 37d9264a4d07e7f5749d561d064ef8c7e32e1d53